File: js/checkout.js
const prompt = require("prompt-sync")();



function unitCheck(unit){

	if(isNaN(Number(unit))) return false;
	else
	if(Number.isInteger(Number(unit))) return true;
	else return false

}


function priceCheck(price){

	if(isNaN(Number(price))) return false
	else return true

}


function discountCheck(discount){

	if(isNaN(Number(discount))) return false

	discount = Number(discount)
	if(discount > 0 && discount < 100) return true
	else return false

}


function paymentCheck(payment, bill){

	if(isNaN(Number(payment))) return false

	payment = Number(payment)
	if(payment < 0 || payment < bill) return false
	else return true

}




function invoice(catalogue, customerName, cashierName, discount){

	const date = new Date().toLocaleString();

	let bill = 0;
	for(const item of catalogue){
		bill += item[3];
	}

	discount = Number(discount);
	let billDiscount = Number(bill * (discount / 100));
	let VATDiscount = Number(bill * 0.175);
	let newBill = Number(bill - (billDiscount + VATDiscount);

	console.log(`

	Al-Qaeda Auctions
	MAIN BRANCH
	LOCATION: secret
	TEL: those who need to know, know
	Date: ${date}
	Cashier: ${cashierName}
	Customer's Name: ${customerName}

	=======================================================
		ITEM	QTY	PRICE(£)	TOTAL(£)
	--------------------------------------------------------
	`);

	for(const item of catalogue){
		console.log(`\t\t${item[0]}\t${item[1]}\t${item[2].toFixed(2)}\t\t${item[3].toFixed(2)}\n`)
	}

	console.log(`
	--------------------------------------------------------
				Sub Total:	${bill.toFixed(2)}
				Discount:	${billDiscount.toFixed(2)}
				VAT @17.5%:	${VATDiscount.toFixed(2)}
	=======================================================
				Bill Total:	${newBill.toFixed(2)}
	=======================================================
	THIS IS NOT A RECEIPT KINDLY PAY ${newBill.toFixed(2)}
	=======================================================

	`);

	return newBill;

}



function receipt(catalogue, customerName, cashierName, discount, payment){

	const date = new Date().toLocaleString();

	let bill = 0;
	for(const item of catalogue){
		bill += item[3];
	}

	discount = Number(discount)
	payment = Number(payment)

	let billDiscount = Number(bill * (discount / 100));
	let VATDiscount = Number(bill * 0.175);
	let newBill = Number(bill - (billDiscount + VATDiscount);
	let balance = Number(payment - newBill);
	

	console.log(`

	Al-Qaeda Auctions
	MAIN BRANCH
	LOCATION: secret
	TEL: those who need to know, know
	Date: ${date}
	Cashier: ${cashierName}
	Customer's Name: ${customerName}

	=======================================================
		ITEM	QTY	PRICE(£)	TOTAL(£)
	--------------------------------------------------------
	`);

	for(const item of catalogue){
		console.log(`\t\t${item[0]}\t${item[1]}\t${item[2].toFixed(2)}\t\t${item[3].toFixed(2)}\n`)
	}

	console.log(`
	--------------------------------------------------------
				Sub Total:	${bill.toFixed(2)}
				Discount:	${billDiscount.toFixed(2)}
				VAT @17.5%:	${VATDiscount.toFixed(2)}
	=======================================================
				Bill Total:	${newBill.toFixed(2)}
				Payment:	${payment.toFixed(2)}
				Balance:	${balance.toFixed(2)}
	=======================================================
					HAPPY BOMBING
	=======================================================


	`);

}





const catalogue = [];
let proceed = true;


while(true){

	let customerName = prompt("What is the customer's name: ");

	while(proceed){
		let item = prompt("What did the user buy? ");

		let units = prompt("how many units were purchased? ");
		if(!unitCheck(units)){
			console.log("Invalid");
			continue;
		}

		let price = prompt("how much per unit: ")
		if(!priceCheck(price)){
			console.log("Invalid");
			continue;
		}


		units = Number(units);
		price = Number(price);
		let total = units * price;

		catalogue.push([item, units, price, total]);

		
		while(proceed){
			let choice = prompt("Enter yes to add more items, no to get invoice: ");
			choice = choice.toUpperCase();

			switch(choice){
				case "YES": break;

				case "NO": {
					proceed = false;
					break;
				}

				default: print("invalid choice, try again");

			}
		} 
	}	

	while(true){
		let cashierName = prompt("What is your name: ");

		let discount = prompt("what percentage discount is being offered? ")
		if(!discountCheck(discount)){
			console.log("Invalid");
			continue;
		}

		let bill = invoice(catalogue, customerName, cashierName, discount);

		while(true){
			let payment = prompt("Your Payment: ");
			if(!paymentCheck(payment, bill)){
				console.log("Invalid");
				continue;
			}

			receipt(catalogue, customerName, cashierName, discount, payment);
			break;
		}
		break;

	}
	break;

}

File: oldstuff/py/checkout.py
import datetime


def name_check(name):

	if not name or name.isspace():
		return "Name cannot be empty"

	if name.startswith(" ") or name.endswith(" "):
		return "Invalid space present"

	new_name = name.replace("-", "").replace(" ", "")

	if len(new_name) < 2 or len(new_name) > 30:
		return "Character limit error"

	if any(char == " " for char in name[:2]):
		return "Invalid name"

	if new_name.isalpha():
		return True
	else:
		return "Invalid name"



def item_check(item):

	if not item or item.isspace():
		return "Space cannot be empty"

	if item.startswith(" ") or item.endswith(" "):
		return "Invalid space present"

	temp = item.replace(" ","")

	if len(temp) < 2 or len(temp) > 30:
		return "Character limit error"

	if any(char == " " for char in item[:2]):
		return "Invalid item"

	if temp.isalnum():
		return True
	else:
		return "Invalid character"





def customer_details(catalogue):

	while True:
		item = str(input("What did the user buy? "))
		if item_check(item) != True:
			print(item_check(item))
			continue
		break

	while True:
		try:
			units = int(input("how many units were purchased? "))
			if units < 1: raise ValueError("units purchased must be positive")

			price = float(input("how much per unit: "))
			if price < 1: raise ValueError("price must be positive")
			if price != round(price, 2): raise ValueError("invalid price")


		except ValueError as e:
			print(e)
			continue


		total = round(units * price, 2)

		catalogue.append([item, units, price, total])
		break


def cashier_details():
	
	while True:
		cash_name = str(input("What is your name: "))
		if name_check(cash_name) != True:
			print(name_check(cash_name))
			continue
		break

	while True:
		try:
			discount = float(input("what percentage discount is being offered? "))

			if discount < 0 or discount > 100: raise ValueError("invalid discount")

		except ValueError as e:
			print(e)
			continue

		return (cash_name, discount)
		


def customer_payment(cost):
	while True:
		try:
			payment = float(input("Your Payment: "))
		
			if payment < 0: raise ValueError("payment cannot be negative")
			if payment < cost: raise ValueError("This will not cover the bill")

		except ValueError as e:
			print(e)
			continue

		return payment




def product_list(catalogue):

	item_width = max(len(purchase[0]) for purchase in catalogue)
	unit_width = max(len(str(purchase[1])) for purchase in catalogue)
	price_width = max(len(str(purchase[2])) for purchase in catalogue)
	total_width = max(len(str(purchase[3])) for purchase in catalogue)

	for purchase in catalogue:
		print(f"\t\t{purchase[0]:>{item_width}}\t{purchase[1]:>{unit_width}}\t{purchase[2]:>{price_width}.2f}\t\t{purchase[3]:>{total_width}.2f}\n")




def invoice(catalogue, customer_name, cashier_info):

	cashier_name, discount = cashier_info

	today = datetime.datetime.now()
	date = today.strftime("%d-%b-%y %-I:%M:%S%p")

	bill = 0
	for purchase in catalogue:
		bill += purchase[3]

	bill_discount = round(bill * (discount / 100), 2)
	VAT_discount = round(bill * 0.175, 2)
	new_bill = round(bill - (bill_discount + VAT_discount), 2)

	

	print(f"""

	Al-Qaeda Auctions
	MAIN BRANCH
	LOCATION: secret
	TEL: those who need to know, know
	Date: {date}
	Cashier: {cashier_name}
	Customer's Name: {customer_name}

	=======================================================
		ITEM	QTY	PRICE(£)	TOTAL(£)
	-------------------------------------------------------
	""")

	product_list(catalogue)

	print(f"""
	-------------------------------------------------------
				Sub Total:	{bill:>7.2f}
				Discount:	{bill_discount:>7.2f}
				VAT @17.5%:	{VAT_discount:>7.2f}
	=======================================================
				Bill Total:	{new_bill:>7.2f}
	=======================================================
		THIS IS NOT A RECEIPT KINDLY PAY {new_bill:.2f}
	=======================================================

	""")

	return new_bill




def receipt(catalogue, customer_name, cashier_info, payment):

	cashier_name, discount = cashier_info

	today = datetime.datetime.now()
	date = today.strftime("%d-%b-%y %-I:%M:%S%p")

	bill = 0
	for purchase in catalogue:
		bill += purchase[3]

	bill_discount = round(bill * (discount / 100), 2)
	VAT_discount = round(bill * 0.175, 2)
	new_bill = round(bill - (bill_discount + VAT_discount), 2)
	balance = round(payment - new_bill, 2)


	print(f"""

	Al-Qaeda Auctions
	MAIN BRANCH
	LOCATION: secret
	TEL: those who need to know, know
	Date: {date}
	Cashier: {cashier_name}
	Customer's Name: {customer_name}

	=======================================================

		ITEM	QTY	PRICE(£)	TOTAL(£)
	--------------------------------------------------------
	""")
	
	product_list(catalogue)
	print(f"""
	--------------------------------------------------------
				Sub Total:	{bill:>7.2f}
				Discount:	{bill_discount:>7.2f}
				VAT @17.5%:	{VAT_discount:>7.2f}
	=======================================================
				Bill Total:	{new_bill:>7.2f}
				Payment:	{payment:>7.2f}
				Balance:	{balance:>7.2f}
	=======================================================
				HAPPY BOMBING
	=======================================================

	""")

	return balance





catalogue = []
proceed = True

while True:
	customer_name = str(input("What is the customer's name: "))

	if name_check(customer_name) != True:
		print(name_check(customer_name))
		continue

	while proceed:
		customer_details(catalogue)

		while proceed:
			choice = str(input("Enter yes to add more items, no to get invoice: "))
			choice = choice.upper()

			match choice:
				case "YES":
					break

				case "NO":
					proceed = False
					break

				case _:
					print("invalid choice, try again")
					continue 


	cashier_info = cashier_details()

	cost = invoice(catalogue, customer_name, cashier_info)
	
	payment = customer_payment(cost)

	receipt(catalogue, customer_name, cashier_info, payment)
	break




#fix alignment issues for all variables in invoice & receipt functions
#round all values to 2 d.p in invoice & receipt functions



